### PR TITLE
Add a tifffile plugin and make it the default for tiff images

### LINKED ISSUE #30: Plugin: scientific TIFF files
In some fields a variant of the TIFF format is heavily used, e.g. for multiplanar microscopic images. Like OME-TIFF (https://www.openmicroscopy.org/site/support/ome-model/ome-tiff/).
The freeimage TIFF reader only deals with basic TIFF.

This is not really my field, so could use some input on what is needed.

One implementation that we could potentially use is: http://www.lfd.uci.edu/~gohlke/code/tifffile.py.html
I believe there was another, but I cannot remember.

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #30.

### CODE REVIEW

File: imageio/plugins/tifffile.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+from .. import formats
+from ..core import Format
+
+import numpy as np
+
+try:
+    import tifffile as _tifffile
+except ImportError:
+    import _tifffile
```
Comment (almarklein) [inline imageio/plugins/tifffile.py]: @blink1073 are you still using Python 2? :P

**Comment by almarklein** [comment]
This is awesome. So if I understand correctly this is Gohlke's tiffile module with improvements by yourself?

**Comment by blink1073** [comment]
I removed the dependency on `importlib` and made it into a PyPI package, but the hard work is all Christophe.

File: imageio/plugins/tifffile.py (hunk 1 above)
Comment (blink1073) [inline imageio/plugins/tifffile.py]: Oh, ha, I have the system `tifffile` installed so it never got to that line.  (I use 3.4)

**Comment by blink1073** [comment]
The only failing test is due to the fact that I added a rather large file.

File: imageio/plugins/tifffile.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        if request.filename.lower().endswith(TIFF_FORMATS):
+            return True  # We support any kind of image data
+        else:
+            return False
+
+    def _can_write(self, request):
+        if request.filename.lower().endswith(TIFF_FORMATS):
+            return True  # We support any kind of image data
+        else:
+            return False
```
Comment (almarklein) [inline imageio/plugins/tifffile.py]: `return request.filename.lower().endswith(TIFF_FORMATS)` ?

File: imageio/plugins/tifffile.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    # -- reader
+
+    class Reader(Format.Reader):
+
+        def _open(self):
+            if not tifffile:
+                load_tifffile()
+            self._tf = tifffile.TiffFile(self.request.get_local_filename())
```
Comment (almarklein) [inline imageio/plugins/tifffile.py]: Does `TiffFile` also accept a file object? It should be faster in some cases. http://imageio.readthedocs.org/en/latest/devapi.html#imageio.core.Request.get_file

File: imageio/plugins/tifffile.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+    class Writer(Format.Writer):
+
+        def _open(self, bigtiff=None, byteorder=None, software=None,
+                  **kwargs):
+            self._parameters = kwargs
+            if not tifffile:
+                load_tifffile()
+            self._tf = tifffile.TiffWriter(self.request.filename,
+                                           bigtiff, byteorder, software)
```
Comment (almarklein) [inline imageio/plugins/tifffile.py]: This should be `request.get_local_filename()` or `request.get_file()`. `request.filename` can also be an http address or zip file.

**Comment by blink1073** [comment]
@almarklein, I addressed your concerns.  The writer would not accept the file provided by `get_file()`, so I used `get_local_filename()`.

File: imageio/plugins/tifffile.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+TIFF_FORMATS = ('.tif', '.tiff', '.stk', '.lsm')
+WRITE_METADATA_KEYS = ('photometric', 'planarconfig', 'resolution',
+                       'description', 'compress', 'volume', 'writeshape',
+                       'extratags')
+READ_METADATA_KEYS = ('planar_configuration', 'is_fluoview', 'is_nih',
+                      'is_contiguous', 'is_micromanager', 'is_ome',
+                      'is_palette', 'is_reduced', 'is_rgb', 'is_sgi',
+                      'is_shaped', 'is_stk', 'is_tiled',
+                      'resolution_unit', 'compression',
+                      'fill_order', 'orientation', 'predictor')
```
Comment (almarklein) [inline imageio/plugins/tifffile.py]: Should the existence of these metadata keys be documented somewhere?

Comment (blink1073) [inline imageio/plugins/tifffile.py]: How about `Metadata on reading` and `Metadata for Writing` sections in the class docstring?

Comment (almarklein) [inline imageio/plugins/tifffile.py]: Yes, that sounds good.

**Comment by blink1073** [comment]
Better?

**Comment by almarklein** [comment]
Looking good! Ready?

**Comment by blink1073** [comment]
Merge away!

**Comment by almarklein** [comment]
Awesome! Beat that PIL! :P

**Comment by kuchi** [comment]
@blink1073 you are awesome! Amazing fast! Wish I could go to spicy 2015 and see your talk.

**Comment by blink1073** [comment]
I am a terrible speaker, I only speak in code :smile:

**Pull request merged**